### PR TITLE
7685.master

### PR DESCRIPTION
**Pull request by ivandudardb**
Implemented the ability to edit information in the block via the administrative menu. 
Configured validation for the entered data.

### CODE REVIEW

File: web/modules/custom/custom_weather/src/Form/TestForm.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+namespace Drupal\custom_weather\Form;
+
+use Drupal\Core\Form\ConfigFormBase;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * {@inheritdoc}
+ */
+class TestForm extends ConfigFormBase {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Form/TestForm.php:11]: There must be a comment on the class

File: web/modules/custom/custom_weather/src/Form/TestForm.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    // Get the value of the selected city from the form.
+    $selected_city = $form_state->getValue('selected_city');
+    // Get the value of the API key from the form.
+    $api_key = $form_state->getValue('api_key');
+
+    // Check if there were validation errors.
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Form/TestForm.php:91]: No verification required

File: web/modules/custom/custom_weather/src/Form/TestForm.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      '#default_value' => $this->config('custom_weather.settings')->get('api_key'),
+    ];
+
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Form/TestForm.php:69]: Add return type to method

File: web/modules/custom/custom_weather/src/Form/TestForm.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    elseif (strlen($api_key) !== 32) {
+      $form_state->setErrorByName('api_key', $this->t('API Key should be 32 characters long.'));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Form/TestForm.php:85]: Add return type to method

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php [hunk 5]
```diff
@@ -14,36 +18,82 @@
...
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The configuration factory.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, ConfigFactoryInterface $config_factory) {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:44]: It`s better to use Constructor Promotion

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php [hunk 6]
```diff
@@ -14,36 +18,82 @@
...
+      $plugin_definition,
+      $container->get('config.factory')
+    );
+  }
 
   /**
    * {@inheritdoc}
    */
-  public function build():array {
+  public function build() {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:66]: Add return type to method

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php [hunk 7]
```diff
@@ -14,36 +18,82 @@
...
     $weatherData = $this->getWeatherApi();
-    $temp = $weatherData['main']['temp'];
+    if (!isset($weatherData['main']['temp']) || !isset($weatherData['weather'][0]['main'])) {
```
Comment (KuzyaWkk) [inline web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:68]: Is this check correct?

File: web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig [hunk 8]
```diff
@@ -25,7 +25,9 @@
  * @see template_preprocess_block()
  */
 #}
- {% set weather_text = content['#weather_text'] %}
+{% set city = content['#selected_city'] %}
```
Comment (KuzyaWkk) [inline web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig:28]: You can use these changes anyway, without the content['*']

File: web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig [hunk 9]
```diff
@@ -51,11 +53,11 @@
      %}
    </div>
     <p class="weather_content">
-      {{ content['#temp'] }}
+      {{ temp }}
     </p>
     <span class="celsium">
       c
```
Comment (KuzyaWkk) [inline web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig:59]: Don't understand the line C? Or the temperature or the letter, what is it?

**Review by KuzyaWkk — CHANGES_REQUESTED** [review]
Please note the comments above

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php (hunk 7 above)
Comment (ivandudardb) [reply web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:57]: Sure

File: web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig (hunk 8 above)
Comment (ivandudardb) [reply web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig:28]: I can't. This is the only possible way.

File: web/modules/custom/custom_weather/custom_weather.routing.yml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+test_form.admin_settings:
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/custom_weather.routing.yml:1]: prefix such names with your module name

File: web/modules/custom/custom_weather/src/Form/TestForm.php [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * Validates the form submission.
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state):void {
+    $api_key = $form_state->getValue('api_key');
+    if (empty($api_key)) {
+      $form_state->setErrorByName('api_key', $this->t('API Key field is required.'));
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Form/TestForm.php:72]: can be handled by #required => TRUE on form element

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php [hunk 12]
```diff
@@ -14,36 +18,73 @@
...
+    );
+  }
 
   /**
    * {@inheritdoc}
    */
   public function build():array {
     $weatherData = $this->getWeatherApi();
-    $temp = $weatherData['main']['temp'];
+    if (!isset($weatherData['main']['temp']) || !isset($weatherData['weather'][0]['main'])) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:59]: is is not logical check, adjust your function

File: web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig (hunk 9 above)
Comment (lavrenchuk-dev) [reply web/themes/custom/customtheme/templates/block/block--customtheme-customweatherblock.html.twig:59]: Better to use some glyph for C

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php [hunk 13]
```diff
@@ -33,7 +33,7 @@ class CustomWeather extends BlockBase implements ContainerFactoryPluginInterface
   /**
    * {@inheritdoc}
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, protected ConfigFactoryInterface $configFactory) {
+  public function __construct(protected ClientFactory $httpClient, array $configuration, $plugin_id, $plugin_definition, protected ConfigFactoryInterface $configFactory) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:36]: it is very weird to have your parameter in construct first, then all parameters as in parent, and then again your parameters
Please account for that in future.

**Review by lavrenchuk-dev — APPROVED** [review]

File: web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php (hunk 13 above)
Comment (lavrenchuk-dev) [reply web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php:36]: Also
you could format a long parameters list like that 
image

**Pull request merged**